### PR TITLE
Pending changes to master

### PR DESCRIPTION
**Pull request by serg7777777serg**
Please review this shit code ))

### CODE REVIEW

File: Chapters/Common/StringAlgorithms.cs [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+            }
+
+        }
+
+        //1
+        public int PunctuationCount()
+        {
+            int sum = 0;
+            foreach (var s in Input)
+                if(char.IsPunctuation(s))
```
Comment (onosov) [inline Chapters/Common/StringAlgorithms.cs]: Why do you need this bicycle. Use linq expression

File: Chapters/Common/StringAlgorithms.cs [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+            return sum;
+        }
+
+        //2
+        public int PointExclamationPointCount()
+        {
+            int sum = 0;
+            foreach (var c in Input)
+                if (c == '.' || c == '!')
+                    sum++;
```
Comment (onosov) [inline Chapters/Common/StringAlgorithms.cs]: Why do you need this bicycle. Use linq expression

File: Chapters/Common/StringAlgorithms.cs [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+            return sum;
+        }
+        
+        //3
+        public void RemoveAllCommas()
+        {
+            for (var i = 0; i < Input.Length; i++)
+                if (Input[i] == ',')
+                    Input.Remove(i, 1);
```
Comment (onosov) [inline Chapters/Common/StringAlgorithms.cs]: This is a bug, try to find it. And this is a bicycle.

File: Chapters/Common/StringAlgorithms.cs [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+        }
+        
+        //4
+        public void RemoveSpacesBeforePunctuation()
+        {
+            for (int i = 1; i < Input.Length; i++)
+                if (Input[i] == ' ' && char.IsPunctuation(Input[i - 1]))
+                    Input.Remove(i, 1);
```
Comment (onosov) [inline Chapters/Common/StringAlgorithms.cs]: This is a bug, try to find it. And this is a bicycle.

File: Chapters/Common/StringAlgorithms.cs [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+        }
+        
+        //5
+        public void InsertNumberAfterSpaces(int num)
+        {
+            for (int i = 0; i < Input.Length - 1; i++)
+                if (Input[i] == ' ')
+                    Input.Insert(i + 1, num.ToString());
```
Comment (onosov) [inline Chapters/Common/StringAlgorithms.cs]: This is not a bug, but try to optimize this. There should be redundant iterations, and this is a bicycle.

File: Chapters/Common/StringAlgorithms.cs [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+        }
+        
+        //6
+        public void InsertNumberBeforeExclamation(int num)
+        {
+            for (int i = 1; i < Input.Length; i++)
+                if (Input[i] == '!')
+                    Input.Insert(i, num.ToString());
```
Comment (onosov) [inline Chapters/Common/StringAlgorithms.cs]: Refactor this, it looks like horrible.

File: Chapters/Common/StringAlgorithms.cs [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+                    
+                    byte[] num = new byte[step];
+                    for (int j = start; j < i; j++)
+                        num[i - j] = Convert.ToByte(Input[j]);
+                    for (int j = num.Length - 1; j > 0; j--)
+                        sum += num[j] * (int)Math.Pow(10, num.Length - 1 - j);
+                }
+                
+            }
+            return sum;
```
Comment (onosov) [inline Chapters/Common/StringAlgorithms.cs]: This method looks like very complex, try to simplify this, and delete this redundant flag.

File: Chapters/Common/StringAlgorithms.cs [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+        }
+
+        //8
+        public void RemoveNumbers()
+        {
+            for (int i = 0; i < Input.Length; i++)
+                if (char.IsDigit(Input[i]))
+                    Input.Remove(i, 1);
```
Comment (onosov) [inline Chapters/Common/StringAlgorithms.cs]: This won't wark, fix this.

File: Chapters/Common/StringAlgorithms.cs [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+                    byte[] num = new byte[step];
+                    for (int j = start; j < i; j++)
+                        num[i - j] = Convert.ToByte(Input[j]);
+                    for (int j = num.Length - 1; j > 0; j--)
+                        number += num[j] * (int)Math.Pow(10, num.Length - 1 - j);
+                    if (BinaryFile.IsSimple(number))
+                        sum += number;
+                }
+            }
+            return sum;
```
Comment (onosov) [inline Chapters/Common/StringAlgorithms.cs]: I've seen some similar code, try to make some universal method for these purposes.

File: Chapters/Common/StringAlgorithms.cs [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+                        word[0] = start;
+                        word[1] = temp;
+                    }
+                    isFirst = false;
+                    temp = 0;
+                    start = 0;
+                }
+            }
+
+            Input.Remove(word[0], word[1]);
```
Comment (onosov) [inline Chapters/Common/StringAlgorithms.cs]: Try to simplify this.

File: Chapters/Common/StringAlgorithms.cs [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+                    byte[] num = new byte[step];
+                    for (int j = start; j < i; j++)
+                        num[i - j] = Convert.ToByte(Input[j]);
+                    for (int j = num.Length - 1; j > 0; j--)
+                        number += num[j] * (int)Math.Pow(10, num.Length - 1 - j);
+                    if (BinaryFile.IsSimple(number))
+                        count++;
+                }
+            }
+            return count;
```
Comment (onosov) [inline Chapters/Common/StringAlgorithms.cs]: Try to simplify this.

File: Chapters/Common/StringAlgorithms.cs [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+        }
+
+        //12
+        public void InsertSpaceAfterPunctuation()
+        {
+            for (int i = 0; i < Input.Length; i++)
```
Comment (onosov) [inline Chapters/Common/StringAlgorithms.cs]: This is not a bug, but try to optimize this. There should be redundant iterations, and this is a bicycle.

File: Chapters/Common/StringAlgorithms.cs [hunk 13]
```diff
@@ -0,0 +1,315 @@
...
+                        StringBuilder s = new StringBuilder();
+                        for (int j = start; j < i; j++)
+                            s.Append(Input[j]);
+                        
+                        for (int j = s.Length - 1; j > 0; j--)
+                            res.Append(s[j]);
+                }
+            }
+            Input = res.ToString();
+        }
```
Comment (onosov) [inline Chapters/Common/StringAlgorithms.cs]: Use linq

File: Chapters/Common/StringAlgorithms.cs [hunk 14]
```diff
@@ -0,0 +1,315 @@
...
+                    {
+                        if (temp < len)
+                            len = temp;
+                        isFirst = false;
+                        temp = 0;
+                        start = 0;
+                    }
+                }
+            }
+        }
```
Comment (onosov) [inline Chapters/Common/StringAlgorithms.cs:228]: i've seen similar code.

**Comment by serg7777777serg** [comment]
Review tasks 901-913

**Pull request closed without merging**